File: Pastura/PasturaTests/Engine/ScoringLogic/PrisonersDilemmaLogicTests.swift
import Testing

@testable import Pastura

@Suite(.timeLimit(.minutes(1)))
struct PrisonersDilemmaLogicTests {
  let logic = PrisonersDilemmaLogic()

  @Test func cooperateCooperate() {
    var state = makeState()
    state.pairings = [Pairing(agent1: "A", agent2: "B", action1: "cooperate", action2: "cooperate")]
    let collector = EventCollector()
    logic.calculate(state: &state, emitter: collector.emit)
    #expect(state.scores["A"] == 3)
    #expect(state.scores["B"] == 3)
  }

  @Test func cooperateBetray() {
    var state = makeState()
    state.pairings = [Pairing(agent1: "A", agent2: "B", action1: "cooperate", action2: "betray")]
    let collector = EventCollector()
    logic.calculate(state: &state, emitter: collector.emit)
    #expect(state.scores["A"] == 0)
    #expect(state.scores["B"] == 5)
  }

  @Test func betrayCooperate() {
    var state = makeState()
    state.pairings = [Pairing(agent1: "A", agent2: "B", action1: "betray", action2: "cooperate")]
    let collector = EventCollector()
    logic.calculate(state: &state, emitter: collector.emit)
    #expect(state.scores["A"] == 5)
    #expect(state.scores["B"] == 0)
  }

  @Test func betrayBetray() {
    var state = makeState()
    state.pairings = [Pairing(agent1: "A", agent2: "B", action1: "betray", action2: "betray")]
    let collector = EventCollector()
    logic.calculate(state: &state, emitter: collector.emit)
    #expect(state.scores["A"] == 1)
    #expect(state.scores["B"] == 1)
  }

  @Test func clearsStatePairingsAfterCalc() {
    var state = makeState()
    state.pairings = [Pairing(agent1: "A", agent2: "B", action1: "cooperate", action2: "cooperate")]
    let collector = EventCollector()
    logic.calculate(state: &state, emitter: collector.emit)
    #expect(state.pairings.isEmpty)
  }

  @Test func emitsScoreUpdateEvent() {
    var state = makeState()
    state.pairings = [Pairing(agent1: "A", agent2: "B", action1: "cooperate", action2: "cooperate")]
    let collector = EventCollector()
    logic.calculate(state: &state, emitter: collector.emit)
    let scoreEvents = collector.events.filter {
      if case .scoreUpdate = $0 { return true }
      return false
    }
    #expect(scoreEvents.count == 1)
  }

  private func makeState() -> SimulationState {
    SimulationState(scores: ["A": 0, "B": 0])
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/ScoringLogic/PrisonersDilemmaLogicTests.kt
package com.pastura.engine

import com.pastura.models.Pairing
import com.pastura.models.SimulationEvent
import com.pastura.models.SimulationState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Kotlin sibling of Swift's `PrisonersDilemmaLogicTests`.
 *
 * Ported for the ADR-023 Stage-3 PR-1 score_calc slice (#501).
 */
class PrisonersDilemmaLogicTests {

    private val logic = PrisonersDilemmaLogic()

    private fun makeState() = SimulationState(scores = mapOf("A" to 0, "B" to 0))

    private fun stateWith(action1: String, action2: String) = makeState().copy(
        pairings = listOf(Pairing(agent1 = "A", agent2 = "B", action1 = action1, action2 = action2)),
    )

    @Test
    fun cooperateCooperate() {
        val next = logic.calculate(stateWith("cooperate", "cooperate")) { }
        assertEquals(3, next.scores["A"])
        assertEquals(3, next.scores["B"])
    }

    @Test
    fun cooperateBetray() {
        val next = logic.calculate(stateWith("cooperate", "betray")) { }
        assertEquals(0, next.scores["A"])
        assertEquals(5, next.scores["B"])
    }

    @Test
    fun betrayCooperate() {
        val next = logic.calculate(stateWith("betray", "cooperate")) { }
        assertEquals(5, next.scores["A"])
        assertEquals(0, next.scores["B"])
    }

    @Test
    fun betrayBetray() {
        val next = logic.calculate(stateWith("betray", "betray")) { }
        assertEquals(1, next.scores["A"])
        assertEquals(1, next.scores["B"])
    }

    @Test
    fun clearsStatePairingsAfterCalc() {
        val next = logic.calculate(stateWith("cooperate", "cooperate")) { }
        assertTrue(next.pairings.isEmpty())
    }

    @Test
    fun emitsScoreUpdateEvent() {
        val events = mutableListOf<SimulationEvent>()
        logic.calculate(stateWith("cooperate", "cooperate")) { events += it }
        assertEquals(1, events.filterIsInstance<SimulationEvent.ScoreUpdate>().size)
    }
}
